packages/core/src/context/index.ts

```diff
@@ -0,0 +1,125 @@
+import { CoreSegmentEvent } from '../events/interfaces'
+
+import { v4 as uuid } from '@lukeed/uuid'
+import { dset } from 'dset'
+import { CoreLogger, LogLevel, LogMessage } from '../logger'
+import Stats, { Metric } from '../stats'
+
+export function isContext(v?: object): v is CoreContext {
+  return !!v && typeof v === 'object' && (v instanceof CoreContext || 'id' in v)
+}
+
+export interface SerializedContext {
+  id: string
+  event: CoreSegmentEvent
+  logs: LogMessage[]
+  metrics?: Metric[]
+}
+
+export interface ContextFailedDelivery {
+  reason: unknown
+}
+
+export interface CancelationOptions {
+  retry?: boolean
+  reason?: string
+  type?: string
+}
+
+export class ContextCancelation {
+  retry: boolean
+  type: string
+  reason?: string
+
+  constructor(options: CancelationOptions) {
+    this.retry = options.retry ?? true
+    this.type = options.type ?? 'plugin Error'
+    this.reason = options.reason ?? ''
+  }
+}
+
+export class CoreContext<Event extends CoreSegmentEvent = CoreSegmentEvent> {
```

Comment:
refactored from the browser Context class to make this new "base" class (meant to be used by node and browser) -- adds a generic, improves decoupling by taking Stats and Logger via constructor injection, removes extraneous access modifiers  like the word "public" to make a bit prettier.